File: Python/2_BathroomSecurity_Part2.py
# --- Part Two ---
# You finally arrive at the bathroom (it's a several minute walk from the lobby so visitors can behold the many fancy conference rooms and water coolers on this floor) and go to punch in the code. Much to your bladder's dismay, the keypad is not at all like you imagined it. Instead, you are confronted with the result of hundreds of man-hours of bathroom-keypad-design meetings:

    # 1
  # 2 3 4
# 5 6 7 8 9
  # A B C
    # D
# You still start at "5" and stop when you're at an edge, but given the same instructions as above, the outcome is very different:

# You start at "5" and don't move at all (up and left are both edges), ending at 5.
# Continuing from "5", you move right twice and down three times (through "6", "7", "B", "D", "D"), ending at D.
# Then, from "D", you move five more times (through "D", "B", "C", "C", "B"), ending at B.
# Finally, after five more moves, you end at 3.
# So, given the actual keypad layout, the code would be 5DB3.

# Using the same instructions in your puzzle input, what is the correct bathroom code?

class button(object):
	def __init__(self, x=-2, y=0):
		self.X = x
		self.Y = y
	def right(self):
		if (self.X + abs(self.Y)) < 2:
			self.X += 1
	def left(self):
		if (-self.X + abs(self.Y)) < 2:
			self.X -= 1
	def down(self):
		if (self.Y + abs(self.X)) < 2:
			self.Y += 1
	def up(self):
		if (-self.Y + abs(self.X)) < 2:
			self.Y -= 1
	def move(self, instr):
		# switcher = {
		# "R": self.right(),
		# "L": self.left(),
		# "U": self.up(),
		# "D": self.down() }
		# func = switcher.get(instr, lambda: 0)
		if instr == 'R':
			self.right()
		elif instr == 'L':
			self.left()
		elif instr == 'U':
			self.up()
		elif instr == 'D':
			self.down()
	def number(self):
		if abs(self.Y) == 1:
			num = 7 + self.Y * 4 + self.X
		else:
			num = 7 + self.Y * 3 + self.X	
		return str(format(num,'x'))

with open("2_BathroomSecurity.txt",'r') as file:
	instructions = file.readlines()

Button = button()
solution = ""
	
for stage in instructions:
	for instruction in stage:
		Button.move(instruction)
	solution += Button.number()
print(solution)
		

File: CPP/2_BathroomSecurity_Part2.cpp
#include <string>
#include <fstream>
#include <iostream>
// #include <iomanip>

class Button {
    public:
    Button(int xCoord, int yCoord);
    void right (void);
    void left (void);
    void down (void);
    void up (void);
    void move (char instr);
    std::string number (void);
        
    private:
        int x, y;
};
Button::Button(int xCoord, int yCoord) {
    x = xCoord;
    y = yCoord;
}
void Button::right(void) {
    if((x + abs(y)) < 2) {
        x++;
    }
}
void Button::left(void) {
    if ((-x + abs(y)) < 2) {
        x--;
    }
}
void Button::down(void) {
    if((y + abs(x)) < 2) {
        y++;
    }
}
void Button::up(void) {
    if((-y + abs(x)) < 2) {
        y--;
    }
}
void Button::move(char instr) {
    if(instr == 'R') {     right();}
    else if(instr == 'L') {left(); }
    else if(instr == 'U') {up();   }
    else if(instr == 'D') {down(); }
}
std::string Button::number(void) {
    int num;
    if(abs(y) == 1) {
        num = 7 + 4*y + x;
    }
    else {
        num = 7 + 3*y + x;
    }
    std::cout << std::hex << num;
}

int main() {   
    std::string stage;
    std::ifstream infile ("2_BathroomSecurity.txt");        // opens file
    
    Button button(-2,0);                                    
    
    while (getline(infile, stage)) {                        // stores line in string stage                    
        for (char& instr : stage) {                         // stores char in char instr
            button.move(instr);                             // moves cursor according to instr
        }
        button.number();                                    // displays current button
    }   
    infile.close();
    return 0;
};